### PR TITLE
[Contrib] Add MKL DNN option

### PR DESCRIPTION
**Pull request by icemelon9**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/apache/incubator-tvm/blob/master/CONTRIBUTORS.md#reviewers) by @ them in the pull request thread.

### CODE REVIEW

File: cmake/modules/contrib/BLAS.cmake [hunk 1]
```diff
@@ -55,3 +55,11 @@ elseif(USE_BLAS STREQUAL "none")
...
   message(FATAL_ERROR "Invalid option: USE_BLAS=" ${USE_BLAS})
 endif()
+
+if(USE_MKL_DNN STREQUAL "ON")
+  find_library(BLAS_LIBRARY_MKLDNN dnnl)
+  list(APPEND TVM_RUNTIME_LINKER_LIBS ${BLAS_LIBRARY_MKLDNN})
+  list(APPEND RUNTIME_SRCS ${CBLAS_CONTRIB_SRC})
+  add_definitions(-DUSE_MKL_DNN=1)
+  message(STATUS "Use MKL DNN library " ${BLAS_LIBRARY_MKLDNN})
+endif()
```
Comment (TaoLv) [inline cmake/modules/contrib/BLAS.cmake]: What's the strategy if both `USE_BLAS` AND `USE_MKL_DNN` are set?

Comment (icemelon9) [reply cmake/modules/contrib/BLAS.cmake]: Update the logic here. MKL DNN will only be used when USE_BLAS is not `none`. When both are set, MKL DNN will be only used in the sgemm op as the library has limited support for the BLAS operators. And I find that MKL DNN kernel achieves better performance than MKL in TVM.

File: src/runtime/contrib/cblas/cblas.cc [hunk 2]
```diff
@@ -31,6 +31,9 @@ extern "C" {
 #else
 #include <cblas.h>
 #endif
+#if USE_MKL_DNN == 1
```
Comment (ZhennanQin) [inline src/runtime/contrib/cblas/cblas.cc]: I think code logic here need a small change as well:
```
#if USE_MKL_BLAS == 1
#include <mkl_cblas.h>
#else if USE_MKL_DNN == 1
#include <dnnl.h>
#else
#include <cblas.h>
#endif
```

File: topi/python/topi/x86/dense.py [hunk 3]
```diff
@@ -32,7 +32,7 @@ def _declaration_dense(cfg, data, weight, bias=None, out_dtype=None):
     if "cblas" in target.libs:
         C = cblas.matmul(data, weight, False, True)
         if bias is not None:
-            C = tvm.compute(C.shape, lambda i, j: C[i, j] + bias[j].astype(out_dtype),
+            C = tvm.compute(C.shape, lambda i, j: C[i, j] + bias[j],
```
Comment (minminsun) [inline topi/python/topi/x86/dense.py]: This change doesn't seem to be related to mkl dnn?

File: cmake/modules/contrib/BLAS.cmake (hunk 1 above)
Comment (TaoLv) [reply cmake/modules/contrib/BLAS.cmake]: Another minor nit is to change `USE_MKL_DNN` to either `USE_DNNL` or `USE_MKLDNN`. The first one aligns with the renaming trend of the library. The second one follows the coding convention in MKL-DNN before renaming and in other projects like MXNet.

File: src/runtime/contrib/cblas/cblas.cc (hunk 2 above)
Comment (gasgallo) [reply src/runtime/contrib/cblas/cblas.cc]: @ZhennanQin As @icemelon9 mentioned [here](https://github.com/apache/incubator-tvm/pull/4323#discussion_r345946543), we need both cblas and dnnl because the latter is used for sgemm only.

**Comment by icemelon9** [comment]
@soiferj No. If the USE_BLAS is OFF, we won't compile MKLDNN ops. See here in the cmake
https://github.com/apache/incubator-tvm/pull/4323/files#diff-9e0c77365976363b16f3bdc58f95fb38R59

One proposal is that I remove `(NOT USE_BLAS STREQUAL "none")` in the condition but I won't add ${CBLAS_CONTRIB_SRC} into src list if USE_BLAS is OFF. So if we have conv2d contrib using MKLDNN in the future, we can still include that file in the src list.

How does this sound?

**Comment by soiferj** [comment]
I see, I think I was looking at an old version. I think that solution sounds good.

File: topi/python/topi/x86/dense.py (hunk 3 above)
Comment (icemelon9) [reply topi/python/topi/x86/dense.py]: yes, this is to fix the bug when using cblas library

File: cmake/modules/contrib/BLAS.cmake (hunk 1 above)
Comment (icemelon9) [reply cmake/modules/contrib/BLAS.cmake]: fixed

Comment (TaoLv) [reply cmake/modules/contrib/BLAS.cmake]: Thanks for the quick turnaround.

**Comment by tqchen** [comment]
Thanks @icemelon9 @TaoLv @soiferj !
